services/src/main/java/org/keycloak/services/clientpolicy/condition/ClientRolesCondition.java

```diff
@@ -61,13 +60,10 @@ public ClientPolicyVote applyPolicy(ClientPolicyContext context) throws ClientPo
     private boolean isRolesMatched(ClientModel client) {
         if (client == null) return false;
 
-        List<String> rolesForMatching = getRolesForMatching();
+        Stream<String> rolesForMatching = getRolesStreamForMatching();
         if (rolesForMatching == null) return false;
 
-        client.getRoles().stream().forEach(i -> ClientPolicyLogger.log(logger, "client role = " + i.getName()));
```

Comment:
Maybe it is possible to add the condition like "if (log.isTraceEnabled()" or something like that to save the iterations over the streams in case that logging is not enabled. Maybe very little performance improvement, but it will be probably there?

Reply:
@mposolda Yes, I'll check whether the log level is trace at first. If so, logging the event in order not to call stream regardless of the log level for considering the performance.